src/esp/geo/geo.cpp

```diff
@@ -58,23 +61,66 @@ std::vector<vec2f> convexHull2D(const std::vector<vec2f>& points) {
 
   return hull;
 }
+/**
+ * Assume the aabb is expressed as the center 'c' and the extent 'e'.
+ * each corner X is nothing but a combination from
+ * (c_x +/- e_x, c_y +/- e_y, c_z +/- e_z)
+ *
+ * Denote y = (+/- e_x, +/- e_y, +/- e_z)
+ *
+ * The corner is transformed by:
+ * x = R * x0 + t,
+ * where x_0, x are the coordinates before and after transformation, t is the
+ * translation.
+ *
+ * x = R * (c0 + y0) + t  = (Rc0 + t) + Ry0    eq(1)
+ *
+ * Our Goal is to find the x_max and x_min after the transformation.
+ *
+ * First, determing the x_max:
+ *
+ * In eq(1), Rc0 + t is a constant, which means max{x} iff max{Ry0}
+ *
+ * R looks like:
+ * [R0, R1, R2]
+ * [R3, R4, R5]
+ * [R6, R7, R8]
+ *
+ * We focus on the 1st entry (the 'x' entry) of Ry0:
+ * y_x =<(R0, R1, R2), (+/- e_x, +/- e_y, +/- e_z)>
+ *     =<(+/- R0, +/- R1, +/- R2), (e_x, e_y, e_z)>
+ *
+ * Now, note that e_x, e_y, e_z are all positive values for any non-degenerate
+ * aabb.
+ *
+ * So y_x reaches MAX when +/- R0, +/- R1, +/- R2 are all >=0
+ * that means max{y_x} = <(|R0|, |R1|, |R2|), (e_x, e_y, e_z)>
+ * (|R0| means the absolute value of R0)
+ *
+ * and likewise for y_y, y_z
+ *
+ * The derivation for x_min is similar since same logic applies.
+ */
+Mn::Range3D getTransformedBB(const Mn::Range3D& range,
+                             const Mn::Matrix4& xform) {
+  Mn::Matrix3x3 absRotationScaling = xform.rotationScaling();
+  // compute the absolute value of the rotationScaling part of the original
+  // transformation matrix
+  for (int iRow = 0; iRow < 3; ++iRow) {
+    for (int jCol = 0; jCol < 3; ++jCol) {
+      absRotationScaling[iRow][jCol] =
+          std::fabs(absRotationScaling[iRow][jCol]);
+    }
+  }
```

Comment:
The above can be 

```cpp
auto absRotationScaling = Mn::Matrix3x3::fromVector(Mn::Math::abs(xform.rotationScaling().toVector()));
```

Explanation: the Math lib has abs()/min()/max()... functions for vectors, but not matrices (to keep it maintainable), so the above is the "canonical way" to do it. 

In any case, when you find yourself looping over matrix/vector elements to do some calculation, there's a high chance such operation is already in the API ;)

Reply:
Thank you. I really need it (your advice)! 

You just touched a pain point I am suffering everyday. I am not familiar with the Magnum library like you are. (I am not complaining, just describing a fact.) 

As you can see, the code is not long, but it cost me a couple of hours this afternoon. Most of the time was spent on searching and reading the Magnum docs, understanding your build-in functions, trying to simply the code. But still, you caught me in so many places that I can improve. :-)

Reply:
Anytime, If I'm around and you're thinking "there has to be some easier way", just ping me on Slack, and I can tell you the API to use ;)

Reply:
> Anytime, If I'm around and you're thinking "there has to be some easier way", just ping me on Slack, and I can tell you the API to use ;)

Thanks again. It is very thoughtful.
The only issue is that we have big time zone difference. :-)